Adds Longest Arithmetic Subsequence Implementation

tejaswi0910 opened the pull request:
<!--
Thank you for your contribution!
In order to reduce the number of notifications sent to the maintainers, please:
- create your PR as draft, cf. https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests,
- make sure that all of the CI checks pass,
- mark your PR as ready for review, cf. https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review
-->

Description
This PR implements the Longest Arithmetic Subsequence algorithm. The algorithm identifies the length of the longest arithmetic subsequence from a given array of integers, where a subsequence is defined as arithmetic if the differences between consecutive elements are the same.

Key Changes:
Added _LongestArithmeticSubsequence.java_ under the _dynamicprogramming_ directory.
The implementation uses dynamic programming with a HashMap to store the differences between elements and keep track of subsequences efficiently.



<!-- For completed items, change [ ] to [x] -->

- [ ] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [ ] This pull request is all my own work -- I have not plagiarized it.
- [ ] All filenames are in PascalCase.
- [ ] All functions and variable names follow Java naming conventions.
- [ ] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [ ] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

vil02 reviewed (changes requested):
Please remove the `main` method and add proper junit tests instead. Prefer `ParameterizedTest`s.

tejaswi0910 commented:
> Please remove the `main` method and add proper junit tests instead. Prefer `ParameterizedTest`s.

sure @vil02 , removed the main function and added tests!

vil02 commented on src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    public static int getLongestArithmeticSubsequenceLength(int[] nums) {
+        // If the array is empty or has only one element, return its length.
+        if (nums == null || nums.length == 0) {
```
What do you think about throwing an exception if `nums` is `null`?

Please add a test, with the input being `null`.

tejaswi0910 replied:
agreed, throwing an exception for the null case will be better to handle the invalid input explicitly

vil02 commented on src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:
```diff
@@ -0,0 +1,45 @@
+            return 1;
+        }
+
+        int n = nums.length;
```
```suggestion
        final int n = nums.length;
```
I am not sure if creating this variable is needed - there is not overhead of using `nums.length` directly.

vil02 commented on src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:
```diff
@@ -0,0 +1,45 @@
+        for (int i = 0; i < n; i++) {
+            dp[i] = new HashMap<>();
+            for (int j = 0; j < i; j++) {
+                int diff = nums[i] - nums[j];
```
```suggestion
                final int diff = nums[i] - nums[j];
```

vil02 commented on src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java:
```diff
@@ -0,0 +1,21 @@
+    void testGetLongestArithmeticSubsequenceLength(int[] nums, int expected) {
+        assertEquals(expected, LongestArithmeticSubsequence.getLongestArithmeticSubsequenceLength(nums));
+    }
+
```
```suggestion

        @ParameterizedTest
    @MethodSource("provideTestCases")
    void testGetLongestArithmeticSubsequenceLengthReversedInput(int[] nums, int expected) {
        ArrayUtils.reverse(nums);
        assertEquals(expected, LongestArithmeticSubsequence.getLongestArithmeticSubsequenceLength(nums));
    }

```

Make sure to:
```java
import org.apache.commons.lang3.ArrayUtils;
```

vil02 replied:
Where did the original test go?

tejaswi0910 replied:
Oops! I might have omitted it while adding the suggested code, but I have added it back now!

vil02 commented on src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:
```diff
@@ -0,0 +1,49 @@
+     * @return the length of the longest arithmetic subsequence
+     */
+    public static int getLongestArithmeticSubsequenceLength(int[] nums) {
+        // Throws an exception if nums is null.
```
```suggestion
```

vil02 commented on src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:
```diff
@@ -0,0 +1,49 @@
+            throw new IllegalArgumentException("Input array cannot be null");
+        }
+
+        // If the array is empty or has only one element, return its length.
```
```suggestion
```

vil02 commented on src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private static Stream<Arguments> provideTestCases() {
+        return Stream.of(Arguments.of(new int[] {3, 6, 9, 12, 15}, 5), Arguments.of(new int[] {1, 7, 10, 13, 14, 19}, 4), Arguments.of(new int[] {1, 2, 3, 4}, 4), Arguments.of(new int[] {}, 0), Arguments.of(new int[] {10}, 1), Arguments.of(new int[] {9, 4, 7, 2, 10}, 3));
```
Please add a test case, when the arithmetic sequence is constant, e.g.:
```suggestion
        return Stream.of(Arguments.of(new int[] {3, 6, 9, 12, 15}, 5), Arguments.of(new int[] {1, 7, 10, 13, 14, 19}, 4), Arguments.of(new int[] {1, 2, 3, 4}, 4), Arguments.of(new int[] {}, 0), Arguments.of(new int[] {10}, 1), Arguments.of(new int[] {9, 4, 7, 2, 10}, 3),
            Arguments.of(new int[] {1, 2, 2, 2, 2, 5}, 4));
```

vil02 commented:
Just a small tip: if you want to modify the last commit, you can use `git commit --amend` and then just push your changes with `--force`.


Happy hacktoberfest!

tejaswi0910 commented:
> Just a small tip: if you want to modify the last commit, you can use `git commit --amend` and then just push your changes with `--force`.
> 
> Happy hacktoberfest!

Thanks @vil02, didn't know this....will surely use this in the future!
